File: ParticleSDKTests/ParticleErrorHelperTests.m
//
//  ParticleErrorHelperTests.m
//  ParticleSDKTests
//
//  Created by Raimundas Sakalauskas on 12/07/2018.
//  Copyright © 2018 Ido Kleinman. All rights reserved.
//

#import <XCTest/XCTest.h>
#import "ParticleErrorHelper.h"

@interface ParticleErrorHelperTests : XCTestCase

@end

@implementation ParticleErrorHelperTests

- (void)testErrorMessage {
    NSString *message;

    //error_description should be prioritized against error
    //{ error: 'error1', error_description: 'error2' }
    message = [ParticleErrorHelper getErrorMessage:[self getJSONDictionary:
            @"{ \"error\" : \"error1\", \"error_description\": \"error2\" }"]];
    XCTAssertTrue([message isEqualToString:@"error2"]);

    //{ error: 'error1' }
    message = [ParticleErrorHelper getErrorMessage:[self getJSONDictionary:
                                            @"{ \"error\" : \"error1\" }"]];
    XCTAssertTrue([message isEqualToString:@"error1"]);


    //{ err: 'error1' }
    message = [ParticleErrorHelper getErrorMessage:[self getJSONDictionary:
                                            @"{ \"err\" : \"error1\" }"]];
    XCTAssertTrue([message isEqualToString:@"error1"]);
    
    
    //{ info: 'error1' }
    message = [ParticleErrorHelper getErrorMessage:[self getJSONDictionary:
                                            @"{ \"info\" : \"error1\"  }"]];
    XCTAssertTrue([message isEqualToString:@"error1"]);
    
    
    
    //{ errors: ['error1', 'error2'] }
    message = [ParticleErrorHelper getErrorMessage:[self getJSONDictionary:
                                            @"{ \"errors\" : [ \"error1\", \"error2\" ] }"]];
    XCTAssertTrue([message isEqualToString:@"error1\r\nerror2"]);
    
    

    //{ errors: [ { error: 'error1' }, { error: 'error2' } ] }
    message = [ParticleErrorHelper getErrorMessage:[self getJSONDictionary:
                                            @"{ \"errors\" : [ {\"error\" : \"error1\"}, {\"error\" : \"error2\"} ] }"]];
    XCTAssertTrue([message isEqualToString:@"error1\r\nerror2"]);


    
    //{ errors: [ { error: { status: 'error1' } }, { error: { status: 'error2' } } ] }
    message = [ParticleErrorHelper getErrorMessage:[self getJSONDictionary:
                                            @"{ \"errors\" : [ {\"error\" : { \"status\" : \"error1\"} }, {\"error\" : { \"status\" : \"error2\"} } ] }"]];
    XCTAssertTrue([message isEqualToString:@"error1\r\nerror2"]);
    
    
    
    // { }
    message = [ParticleErrorHelper getErrorMessage:[self getJSONDictionary:@"{ }"]];
    XCTAssertTrue([message isEqualToString:@"Unknown error occurred."]);
}

- (NSDictionary *)getJSONDictionary:(NSString *)jsonString {
    NSData* data = [jsonString dataUsingEncoding:NSUTF8StringEncoding];
    return [NSJSONSerialization JSONObjectWithData:data options:kNilOptions error:nil];
}

@end

File: ParticleSDKTests/Cases/Helpers/ParticleErrorHelperTests.swift
//
//  ParticleErrorHelperTests.swift
//  ParticleSDKTests
//
//  Created by Raimundas Sakalauskas on 13/12/2019.
//  Copyright © 2019 Particle Inc. All rights reserved.
//

import XCTest
@testable import ParticleSDK

class ParticleErrorHelperTests: XCTestCase {

    func test_errorDescriptionSet_PrioritizedOverError() {
        let json = "{ \"error\" : \"error1\", \"error_description\": \"error2\" }"
        let expectedMessage = "error2"
        
        let message = ParticleErrorHelper.getErrorMessage(getJSONDict(withJSONString: json))
        XCTAssertEqual(message, expectedMessage)
    }
  
    func test_errorSet_parsedCorrectly() {
        let json = "{ \"error\" : \"error1\" }"
        let expectedMessage = "error1"
        
        let message = ParticleErrorHelper.getErrorMessage(getJSONDict(withJSONString: json))
        XCTAssertEqual(message, expectedMessage)
    }
    
    func test_errSet_parsedCorrectly() {
        let json = "{ \"err\" : \"error1\" }"
        let expectedMessage = "error1"
        
        let message = ParticleErrorHelper.getErrorMessage(getJSONDict(withJSONString: json))
        XCTAssertEqual(message, expectedMessage)
    }

    func test_infoSet_parsedCorrectly() {
        let json = "{ \"info\" : \"error1\" }"
        let expectedMessage = "error1"
        
        let message = ParticleErrorHelper.getErrorMessage(getJSONDict(withJSONString: json))
        XCTAssertEqual(message, expectedMessage)
    }

    func test_multipleErrorsSet_parsedCorrectly() {
        let json = "{ \"errors\" : [ \"error1\", \"error2\" ] }"
        let expectedMessage = "error1\r\nerror2"
        
        let message = ParticleErrorHelper.getErrorMessage(getJSONDict(withJSONString: json))
        XCTAssertEqual(message, expectedMessage)
    }

    func test_errorDictionarySet_parsedCorrectly() {
        let json = "{ \"errors\" : [ {\"error\" : \"error1\"}, {\"error\" : \"error2\"} ] }"
        let expectedMessage = "error1\r\nerror2"
        
        let message = ParticleErrorHelper.getErrorMessage(getJSONDict(withJSONString: json))
        XCTAssertEqual(message, expectedMessage)
    }
    
    func test_errorAndStatusDictionarySet_parsedCorrectly() {
        let json = "{ \"errors\" : [ {\"error\" : { \"status\" : \"error1\"} }, {\"error\" : { \"status\" : \"error2\"} } ] }"
        let expectedMessage = "error1\r\nerror2"
        
        let message = ParticleErrorHelper.getErrorMessage(getJSONDict(withJSONString: json))
        XCTAssertEqual(message, expectedMessage)
    }
    
    func test_noErrorSet_templateErrorReturned() {
        let json = "{ }"
        let expectedMessage = "Unknown error occurred."
        
        let message = ParticleErrorHelper.getErrorMessage(getJSONDict(withJSONString: json))
        XCTAssertEqual(message, expectedMessage)
    }
    

    func getJSONDict(withJSONString inputString: String) -> [AnyHashable: Any] {
        let data = inputString.data(using: .utf8)!
        return try! JSONSerialization.jsonObject(with: data, options: []) as! [AnyHashable: Any]
    }
}
